flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierAligner.java

```diff
@@ -46,14 +49,8 @@
 
 	private static final Logger LOG = LoggerFactory.getLogger(CheckpointBarrierAligner.class);
 
-	/** Used to get InputGate by channel index. */
-	private final InputGate[] channelIndexToInputGate;
-
-	/** Used to get channel index offset by InputGate. */
-	private final Map<InputGate, Integer> inputGateToChannelIndexOffset;
-
 	/** Flags that indicate whether a channel is currently blocked/buffered. */
-	private final boolean[] blockedChannels;
+	private final Map<InputChannelInfo, Boolean> blockedChannels;
```

Comment:
Why do we use `Map<Key, Boolean>` instead of just  `Set<Key>` (also in `CheckpointBarrierUnaligner`, `ThreadSafeUnaligner`)?

(I guess we can avoid the penalty of dynamic reallocation by setting set capacity in advance)

Reply:
Yes, you are right (I tried to explain it in the commit message). Also internally `HashSet` uses `LinkedHashMap`, so it's even then slower than using `HashMap` directly even with dynamic reallocation.

Reply:
> Also internally `HashSet` uses `LinkedHashMap`

Can you explain what do you mean?
In my OpenJDK 11 `HashSet` uses `HashMap` under the hood, and I believe this is what most implementations do.

Reply:
Never mind, I misread the code. It's only used while calling LinkedHashSet -> super(HashSet).

Anyways, the current way is used to avoid dynamic reallocations. I can also write a small wrapper Set implementation (similar to EnumSet) if it's too hard to read right now.

Reply:
I'm not sure if it's not a premature optimization (I think contains is faster for set but put/remove slower; and it all may not be visible).

But I'm also fine with the current approach.

Reply:
`HashSet#contains` uses `HashMap#containsKey`, which fetches the node of the key. Without hash collisions and overflow lists, that is exactly the same. The question is how many collisions we have. It may also depend on whether the respective `Set` is mostly filled or empty (e.g., `blockedChannels` should be mostly empty). We might want to potentially invert the current semantics to keep them empty as long as possible.

My main motivation was to avoid adding anything to GC pressure and using mostly a constant data structure as before to not change too much. I think I'd leave it as is for now. We can revise it when we adjust the threading model of Unaligner, which will be a major change on the code anyways.